# 인구 데이터 기반 소득 예측 경진대회

이번 대회는 인구 데이터 바탕으로 소득이 5만달러 이하인지 초과인지 분류하는 대회입니다.

간단한 모델을 작성해서 성능을 확인하고, 성능을 개선하기 위해서는 어떻게 접근해야 하는지 생각해봅시다!

그럼 시작해볼까요?

## 데이터 불러오기

먼저 분석하려는 데이터를 작업장으로 가져오는 작업이 필요합니다.

이를 위해서 파이썬 라이브러리 중 하나인 Pandas를 이용합니다.

pandas 라이브러리는 엑셀과 같은 행과 열로 이루어진 테이블(table) 형식의 데이터를 편하게 다루도록 해주는 라이브러리입니다.

In [3]:
import pandas as pd

# csv 형식으로 된 데이터 파일을 읽어옵니다.
train = pd.read_csv('data/train.csv')
train

id  age workclass  fnlwgt     education  education.num  \
0          0   32   Private  309513    Assoc-acdm             12   
1          1   33   Private  205469  Some-college             10   
2          2   46   Private  149949  Some-college             10   
3          3   23   Private  193090     Bachelors             13   
4          4   55   Private   60193       HS-grad              9   
...      ...  ...       ...     ...           ...            ...   
17475  17475   35       NaN  320084     Bachelors             13   
17476  17476   30       NaN   33811     Bachelors             13   
17477  17477   71       NaN  287372     Doctorate             16   
17478  17478   41       NaN  202822       HS-grad              9   
17479  17479   72       NaN  129912       HS-grad              9   

           marital.status       occupation   relationship                race  \
0      Married-civ-spouse     Craft-repair        Husband               White   
1      Married-civ-spouse  Exec-managerial        Husband               White   
2      Married-civ-spouse     Craft-repair        Husband               White   
3           Never-married     Adm-clerical      Own-child               White   
4                Divorced     Adm-clerical  Not-in-family               White   
...                   ...              ...            ...                 ...   
17475  Married-civ-spouse              NaN           Wife               White   
17476       Never-married              NaN  Not-in-family  Asian-Pac-Islander   
17477  Married-civ-spouse              NaN        Husband               White   
17478           Separated              NaN  Not-in-family               Black   
17479  Married-civ-spouse              NaN        Husband               White   

          sex  capital.gain  capital.loss  hours.per.week native.country  \
0        Male             0             0              40  United-States   
1        Male             0             0              40  United-States   
2        Male             0             0              40  United-States   
3      Female             0             0              30  United-States   
4      Female             0             0              40  United-States   
...       ...           ...           ...             ...            ...   
17475  Female             0             0              55  United-States   
17476  Female             0             0              99  United-States   
17477    Male             0             0              10  United-States   
17478  Female             0             0              32  United-States   
17479    Male             0             0              25  United-States   

       target  
0           0  
1           1  
2           0  
3           0  
4           0  
...       ...  
17475       1  
17476       0  
17477       1  
17478       0  
17479       0  

[17480 rows x 16 columns]

In [4]:
train.shape

(17480, 16)

총 17480행과 16열을 가진 데이터 입니다.

## 결측치 확인

결측치(NA: Not Available)란 값이 누락된 데이터를 말합니다.

보다 정확한 분석을 하기 위해서는 데이터의 결측치를 확인하고 적절히 처리해주어야 합니다.

이번 데이터에 결측치가 있나 확인해볼까요?

In [14]:
def check_missing_col(dataframe):
    missing_col = []
    for col in dataframe.columns:
        missing_values = sum(dataframe[col].isna())
        is_missing = True if missing_values >= 1 else False
        if is_missing:
            print(f'결측치가 있는 컬럼은: {col} 입니다')
            print(f'해당 컬럼에 총 {missing_values} 개의 결측치가 존재합니다.')
            missing_col.append([col, dataframe[col].dtype])
    if missing_col == []:
        print('결측치가 존재하지 않습니다')
    return missing_col

missing_col = check_missing_col(train)

결측치가 있는 컬럼은: workclass 입니다
해당 컬럼에 총 1836 개의 결측치가 존재합니다.
결측치가 있는 컬럼은: occupation 입니다
해당 컬럼에 총 1843 개의 결측치가 존재합니다.
결측치가 있는 컬럼은: native.country 입니다
해당 컬럼에 총 583 개의 결측치가 존재합니다.


결측치가 존재하는군요!

결측치 데이터가 범주형인지 수치형인지 unique() 메소드를 통하여 확인하겠습니다.

In [15]:
print(train['workclass'].unique())
print(train['occupation'].unique())
print(train['native.country'].unique())

['Private' 'State-gov' 'Local-gov' 'Self-emp-not-inc' 'Self-emp-inc'
 'Federal-gov' 'Without-pay' nan 'Never-worked']
['Craft-repair' 'Exec-managerial' 'Adm-clerical' 'Prof-specialty'
 'Machine-op-inspct' 'Other-service' 'Sales' 'Farming-fishing'
 'Transport-moving' 'Handlers-cleaners' 'Tech-support' 'Protective-serv'
 'Priv-house-serv' 'Armed-Forces' nan]
['United-States' 'Poland' 'Mexico' 'Ireland' 'Guatemala'
 'Dominican-Republic' 'Greece' 'El-Salvador' 'Portugal' 'Canada'
 'Philippines' 'India' 'Italy' 'England' 'Jamaica' 'Columbia' 'South'
 'Vietnam' 'Cuba' 'Laos' 'Hong' 'Haiti' 'Germany' 'Yugoslavia' 'Ecuador'
 'France' 'Puerto-Rico' 'Outlying-US(Guam-USVI-etc)' 'Taiwan' 'China'
 'Japan' 'Honduras' 'Peru' 'Nicaragua' 'Hungary' 'Cambodia' 'Iran'
 'Trinadad&Tobago' 'Thailand' 'Scotland' 'Holand-Netherlands' nan]


모두 데이터 값이 범주형이고   
결측치들에 대해서 어떤 특별한 패턴은 보이지 않으므로  
범주형 데이터에 대해서는 행을 삭제해주겠습니다.

In [16]:
# 결측치를 처리하는 함수를 작성합니다.
def handle_na(data, missing_col):
    temp = data.copy()
    for col, dtype in missing_col:
        if dtype == 'O':
            # 범주형 feature가 결측치인 경우 해당 행들을 삭제해 주었습니다.
            temp = temp.dropna(subset=[col])
    return temp

train = handle_na(train, missing_col)

# 결측치 처리가 잘 되었는지 확인해 줍니다.
missing_col = check_missing_col(train) 

결측치가 존재하지 않습니다


## 데이터 전처리

info() 메소드를 이용하여 데이터의 타입을 보면 상당히 많은 데이터가 범주형인 것을 확인 할 수 있습니다.


In [5]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17480 entries, 0 to 17479
Data columns (total 16 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   id              17480 non-null  int64 
 1   age             17480 non-null  int64 
 2   workclass       15644 non-null  object
 3   fnlwgt          17480 non-null  int64 
 4   education       17480 non-null  object
 5   education.num   17480 non-null  int64 
 6   marital.status  17480 non-null  object
 7   occupation      15637 non-null  object
 8   relationship    17480 non-null  object
 9   race            17480 non-null  object
 10  sex             17480 non-null  object
 11  capital.gain    17480 non-null  int64 
 12  capital.loss    17480 non-null  int64 
 13  hours.per.week  17480 non-null  int64 
 14  native.country  16897 non-null  object
 15  target          17480 non-null  int64 
dtypes: int64(8), object(8)
memory usage: 2.1+ MB


컴퓨터는 숫자만을 인식할 수 있기 때문에, 문자열로 이루어진 값을 숫자로 바꾸어주어야 합니다.

그래서 문자열 형식의 범주형 속성은 모두 숫자 값으로 변환/인코딩 되어야 합니다.

이번에 적용해볼 인코딩 방식은 'Label Encoding' 입니다.

Label Encoding이란, 범주형 변수의 문자열을 수치형으로 변환하는 방법 중 하나입니다.

Label Encoding을 통해 숫자형으로 표현해보겠습니다.

In [18]:
#라벨인코딩을 하기 위함 dictionary map 생성 함수
def make_label_map(dataframe):
    label_maps = {}
    for col in dataframe.columns:
        if dataframe[col].dtype=='object':
            label_map = {'unknown':0}
            for i, key in enumerate(dataframe[col].unique()):
                label_map[key] = i  #새로 등장하는 유니크 값들에 대해 1부터 1씩 증가시켜 키값을 부여해줍니다.
            label_maps[col] = label_map
    return label_maps

# 각 범주형 변수에 인코딩 값을 부여하는 함수
def label_encoder(dataframe, label_map):
    for col in dataframe.columns:
        if dataframe[col].dtype=='object':
            dataframe[col] = dataframe[col].map(label_map[col])
            #dataframe[col] = dataframe[col].fillna(label_map[col]['unknown']) #혹시 모를 결측값은 unknown의 값(0)으로 채워줍니다.
    return dataframe

train = label_encoder(train, make_label_map(train))

In [19]:
train

id  age  workclass  fnlwgt  education  education.num  \
0          0   32          0  309513          0             12   
1          1   33          0  205469          1             10   
2          2   46          0  149949          1             10   
3          3   23          0  193090          2             13   
4          4   55          0   60193          3              9   
...      ...  ...        ...     ...        ...            ...   
15076  15076   35          0  337286          7             14   
15077  15077   36          0  182074          1             10   
15078  15078   50          4  175070          4             15   
15079  15079   39          0  202937          1             10   
15080  15080   33          0   96245          0             12   

       marital.status  occupation  relationship  race  sex  capital.gain  \
0                   0           0             0     0    0             0   
1                   0           1             0     0    0             0   
2                   0           0             0     0    0             0   
3                   1           2             1     0    1             0   
4                   2           2             2     0    1             0   
...               ...         ...           ...   ...  ...           ...   
15076               1           1             2     2    0             0   
15077               2           2             2     0    0             0   
15078               0           3             0     0    0             0   
15079               2          10             2     0    1             0   
15080               0           3             0     0    0             0   

       capital.loss  hours.per.week  native.country  target  
0                 0              40               0       0  
1                 0              40               0       1  
2                 0              40               0       0  
3                 0              30               0       0  
4                 0              40               0       0  
...             ...             ...             ...     ...  
15076             0              40               0       0  
15077             0              45               0       0  
15078             0              45               0       1  
15079             0              40               1       0  
15080             0              50               0       0  

[15081 rows x 16 columns]

다음과 같이 Label Encoding을 통해서 데이터들이 모두 숫자형으로 변환됨을 확인할 수 있습니다.

## 변수 및 모델 정의

우리의 목적은 주어진 와인의 화학적 성분을 이용하여 와인의 품질, 맛 평점이 몇점인지를 예측하는 것입니다.

그럼 인구 데이터가 X, 수입이 y가 되겠죠?

따라서 모델 학습을 위해 학습 데이터와 타겟을 분리해보도록 하겠습니다.

In [21]:
# 분석에 필요 없는 id 와 예측하고자 하는 값 타겟을 제거해줍니다.

X = train.drop(['id', 'target'], axis=1)
y = train['target']

정형데이터에서 가장 익숙한 Linear Regression을 진행해보도록 하겠습니다.

그런데 여기서는 예측하고자 하는 것이 연속적인 값을 갖는 수치 형태가 아닌, 1 단위로 끊어지는 분류 문제이기 때문에

category를 분류할 때 사용하는 LogisticRegression을 이용하여 진행하도록 하겠습니다.

In [22]:
from sklearn.linear_model import LogisticRegression

# 모델 선언
model = LogisticRegression(solver='liblinear')

# 모델 학습
model.fit(X, y)

LogisticRegression(solver='liblinear')

## 모델 성능 확인

굉장히 간단하게 학습이 끝났습니다.

이제 이 모델이 실제로 데이터를 얼마나 잘 맞추는지 확인해볼까요?

In [23]:
# 먼저 점수를 메기는 방법인 평가 지표(Metric)를 정의합니다.
import numpy as np

def ACCURACY(true, pred):   
    score = np.mean(true==pred)
    return score

In [67]:
# 모델의 예측과 실제 정답값을 비교합니다.
prediction = model.predict(X)

score = ACCURACY(y, prediction)

print(f"모델의 정확도는 {score*100:.2f}% 입니다")

모델의 정확도는 79.09% 입니다


## 추론 (Inference)

In [24]:
# csv형식으로 된 데이터 파일을 읽어옵니다.
test = pd.read_csv('data/test.csv')
test.head()

id  age     workclass  fnlwgt     education  education.num  \
0   0   47       Private  304133  Some-college             10   
1   1   34  Self-emp-inc  154227  Some-college             10   
2   2   31     Local-gov  158291     Bachelors             13   
3   3   28       Private  183155     Bachelors             13   
4   4   54     Local-gov  182543  Some-college             10   

       marital.status       occupation   relationship   race     sex  \
0  Married-civ-spouse  Exec-managerial        Husband  White    Male   
1       Never-married            Sales  Not-in-family  White    Male   
2       Never-married     Craft-repair  Not-in-family  White    Male   
3  Married-civ-spouse   Prof-specialty        Husband  White    Male   
4             Widowed     Adm-clerical      Unmarried  White  Female   

   capital.gain  capital.loss  hours.per.week native.country  
0             0             0              45  United-States  
1             0             0              75  United-States  
2          8614             0              40  United-States  
3             0             0              55  United-States  
4             0             0              40         Mexico

마찬가지로 테스트 셋에서도 똑같이 전처리를 해줍니다.

In [70]:
test = label_encoder(test, make_label_map(test))
test = test.drop(['id'],axis=1)
test.head()

age  workclass  fnlwgt  education  education.num  marital.status  \
0   47          1  304133          1             10               1   
1   34          2  154227          1             10               2   
2   31          3  158291          2             13               2   
3   28          1  183155          2             13               1   
4   54          3  182543          1             10               3   

   occupation  relationship  race  sex  capital.gain  capital.loss  \
0           1             1     1    1             0             0   
1           2             2     1    1             0             0   
2           3             2     1    1          8614             0   
3           4             1     1    1             0             0   
4           5             3     1    2             0             0   

   hours.per.week  native.country  
0              45               1  
1              75               1  
2              40               1  
3              55               1  
4              40               2

In [26]:
# 전처리가 완료된 테스트 데이터셋을 통해 본격적으로 학습한 모델로 추론을 시작합니다.
prediction = model.predict(test)
prediction

array([0, 0, 1, ..., 0, 0, 0])

## Dacon 대회에 제출하기 (Submission)

이제 학습한 모델을 사용하여 이 예측 결과를 submission.csv 파일로 만들어서 대회 페이지에 제출해보도록 합시다.

제출한 뒤 리더보드를 통해 결과를 확인합시다.

In [33]:
# 제출용 Sample 파일을 불러옵니다
submission = pd.read_csv('data/sample_submission.csv')
submission.head()

id  target
0   0       0
1   1       0
2   2       0
3   3       0
4   4       0

In [34]:
submission['target'] = prediction

# 데이터가 잘 들어갔는지 확인합니다
submission

id  target
0          0       0
1          1       0
2          2       1
3          3       0
4          4       0
...      ...     ...
15076  15076       1
15077  15077       0
15078  15078       0
15079  15079       0
15080  15080       0

[15081 rows x 2 columns]

submission을 csv 파일로 저장합니다.   
index=False란 추가적인 id를 부여할 필요가 없다는 뜻입니다.    
정확한 채점을 위해 꼭 index=False를 넣어주세요.

In [74]:
submission.to_csv('submit.csv', index=False)

이렇게 생성된 submission.csv 파일을 데이콘 대회 페이지에 업로드 & 제출하여 결과를 확인해보세요!

문제를 해결하기 위한 여러분의 방법을 코드 공유 게시판에 공유해주세요

좋아요와 댓글을 합산하여 가장 높은 점수를 얻으신 분께 데이콘 후드가 제공됩니다!